File: aulas/blockchain/exercicio/progresso.py
#!/usr/bin/env python3
"""Seu painel de progresso. Rode:  python progresso.py

Mostra, parte a parte: se você ligou em config.py ([VOCÊ]) ou ainda usa o
gabarito ([gabarito]), e quantos testes daquela parte passam.
"""

import os
import re
import subprocess
import sys

HERE = os.path.dirname(os.path.abspath(__file__))
sys.path.insert(0, HERE)
from minichain import config  # noqa: E402

PARTS = [
    ("merkle", "Parte 1 · Árvore de Merkle",          "tests/test_merkle.py"),
    ("tx",     "Parte 2 · Transações assinadas",      "tests/test_tx.py"),
    ("block",  "Parte 3 · Cabeçalho & hash do bloco", "tests/test_block.py"),
    ("pow",    "Parte 4 · Mineração (proof of work)", "tests/test_pow.py"),
    ("chain",  "Parte 5 · Encadeamento & validação",  "tests/test_chain.py"),
    ("fork",   "Parte 6 · Cadeia mais longa (forks)", "tests/test_fork.py"),
]


def run(path):
    r = subprocess.run(
        [sys.executable, "-m", "pytest", path, "--no-header",
         "--tb=no", "-p", "no:cacheprovider"],
        cwd=HERE, capture_output=True, text=True,
    )
    out = r.stdout + r.stderr
    passed = int((re.search(r"(\d+) passed", out) or [0, 0])[1])
    failed = int((re.search(r"(\d+) failed", out) or [0, 0])[1])
    errors = int((re.search(r"(\d+) error", out) or [0, 0])[1])
    total = passed + failed + errors
    return passed, total


def main():
    print("\n  minichain — seu progresso")
    print("  " + "─" * 52)
    mine_done = 0
    for key, title, path in PARTS:
        is_mine = config.MINHA[key]
        passed, total = run(path)
        tag = "[VOCÊ]    " if is_mine else "[gabarito]"
        mark = "✓" if passed == total and total else "✗"
        if is_mine and passed == total and total:
            mine_done += 1
        print(f"  {title:<36} {tag}  {passed}/{total} {mark}")
    print("  " + "─" * 52)
    print(f"  Partes suas concluídas: {mine_done}/{len(PARTS)}"
          "   (ligue cada uma em config.py ao terminar)\n")


if __name__ == "__main__":
    main()

File: aulas/blockchain/exercicio-sol/progresso.js
#!/usr/bin/env node
// Seu painel de progresso. Rode:  node progresso.js
//
// Mostra, exercicio a exercicio: se voce ligou em config.js ([VOCE]) ou ainda
// usa o gabarito ([gabarito]), e quantos testes daquele exercicio passam.

const { execFileSync } = require("child_process");
const config = require("./config");

const PARTS = [
  ["cofre",        "1 · Cofre (payable, saque do dono)",   "test/cofre.test.js"],
  ["votacao",      "2 · Votacao (prazo, 1 voto)",          "test/votacao.test.js"],
  ["crowdfunding", "3 · Crowdfunding (meta, refund)",      "test/crowdfunding.test.js"],
  ["leilao",       "4 · Leilao ingles (pull payments)",    "test/leilao.test.js"],
  ["moeda",        "5 · Moeda ERC-20",                     "test/moeda.test.js"],
  ["dvp",          "6 · DvP (entrega contra pagamento)",   "test/dvp.test.js"],
];

function run(file) {
  let out;
  try {
    out = execFileSync("npx", ["hardhat", "test", file], { encoding: "utf8" });
  } catch (e) {
    out = (e.stdout || "") + (e.stderr || "");
  }
  const passed = Number((out.match(/(\d+) passing/) || [0, 0])[1]);
  const failed = Number((out.match(/(\d+) failing/) || [0, 0])[1]);
  return [passed, passed + failed];
}

function main() {
  process.stdout.write("\n  exercicio-sol — seu progresso\n");
  process.stdout.write("  " + "─".repeat(52) + "\n");
  let done = 0;
  for (const [key, title, file] of PARTS) {
    const isMine = config[key];
    const [passed, total] = run(file);
    const tag = isMine ? "[VOCE]    " : "[gabarito]";
    const mark = total && passed === total ? "✓" : "✗";
    if (isMine && total && passed === total) done++;
    process.stdout.write(`  ${title.padEnd(38)} ${tag}  ${passed}/${total} ${mark}\n`);
  }
  process.stdout.write("  " + "─".repeat(52) + "\n");
  process.stdout.write(
    `  Exercicios seus concluidos: ${done}/${PARTS.length}` +
    "   (ligue cada um em config.js ao terminar)\n\n"
  );
}

main();
